rust/src/settings_manager.rs

```diff
@@ -1,72 +1,133 @@
-use std::sync::Mutex;
+use std::fmt;
 
-use yaml_rust::{Yaml, YamlLoader};
+use serde::{
+    de::{self, Unexpected},
+    Deserialize,
+};
 
-const SOURCE: &str = include_str!("../../settings.yaml");
+lazy_static::lazy_static! {
+    static ref SETTINGS: Vec<Setting> = {
+        serde_yaml::from_str(include_str!("../../settings.yaml"))
+            .expect("Could not parse settings.yaml")
+    };
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct Setting {
+    name: String,
+
+    group: String,
+
+    #[serde(rename = "type")]
+    kind: Option<SettingKind>,
+
+    #[serde(deserialize_with = "deserialize_bool", default)]
+    expert: bool,
+
+    #[serde(deserialize_with = "deserialize_bool", default)]
+    readonly: bool,
+
+    #[serde(rename = "Description")]
+    description: Option<String>,
 
-#[derive(PartialEq, Debug)]
-pub enum SettingType {
-    StInteger,
-    StBoolean,
-    StDouble,
-    StString,
-    StEnum,
+    #[serde(alias = "default value")]
+    default_value: Option<String>,
+
+    #[serde(rename = "Notes")]
+    notes: Option<String>,
+
+    units: Option<String>,
+
+    #[serde(rename = "enumerated possible values")]
+    enumerated_possible_values: Option<String>,
+
+    digits: Option<String>,
 }
 
-lazy_static::lazy_static! {
-    static ref SETTINGS: Mutex<Yaml> = {
-        let mut docs = YamlLoader::load_from_str(SOURCE).expect("Could not parse settings.yaml");
-        Mutex::new(docs.remove(0))
-    };
+#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
+pub enum SettingKind {
+    #[serde(rename = "integer", alias = "int")]
+    Integer,
+
+    #[serde(rename = "boolean", alias = "bool")]
+    Boolean,
+
+    #[serde(rename = "float")]
+    Float,
+
+    #[serde(rename = "double", alias = "Double")]
+    Double,
+
+    #[serde(rename = "string")]
+    String,
+
+    #[serde(rename = "enum")]
+    Enum,
+
+    #[serde(rename = "packed bitfield")]
+    PackedBitfield,
```

Comment:
There is only one setting with this type, not sure the best way to handle it, maybe just as a string?

Reply:
String is probably fine

Reply:
(The setting will get written/read as a string anyway so as long as we are able to preserve all the data then a String should be fine)